### DISCUSSION TITLE
Is there a heartbeat check when distcc remotely compiles?

### DISCUSSION THREAD

**Discussion by zcfh** [distcc/distcc#534] [Q&A] [answered]
At present, when some large files and the compilation time is relatively long, it **may be due to network reasons**. After the compilation is completed, there is a probability that the compilation product will fail to return.
```
distccd[34098] (2024-12-16 10:21:19) (dcc_spawn_child) child started as pid38565
distccd[34098] (2024-12-16 10:26:20) (dcc_collect_child) cc child 38565 terminated with status 0
distccd[34098] (2024-12-16 10:26:20) (dcc_collect_child) cc times: user 0.000000s, system 0.000000s, 0 minflt, 0 majflt
distccd[34098] (2024-12-16 10:26:20) (dcc_x_token_int) send DONE00000001
distccd[34098] (2024-12-16 10:26:20) (dcc_x_token_int) send STAT00000000
distccd[34098] (2024-12-16 10:26:20) (dcc_x_file) send 0 byte file /tmp/distcc_33608e9a.stderr with token SERR and compression 69
distccd[34098] (2024-12-16 10:26:20) (dcc_x_token_int) send SERR00000000
distccd[34098] (2024-12-16 10:26:20) (dcc_x_file) send 0 byte file /tmp/distcc_333b8e9a.stdout with token SOUT and compression 69
distccd[34098] (2024-12-16 10:26:20) (dcc_x_token_int) send SOUT00000000
distccd[34098] (2024-12-16 10:26:20) (dcc_x_file) send 201043280 byte file /tmp/distccd_345c8e9a.o with token DOTO and compression 69
distccd[34098] (2024-12-16 10:26:20) (dcc_x_token_int) send DOTO0bfbad50
distccd[34098] (2024-12-16 10:26:20) (dcc_pump_sendfile) ==== 2024-12-16T10:26:20 sendfile: partial transmission of 41880 bytes; retrying 201001400 @41880
distccd[34098] (2024-12-16 10:26:20) (dcc_pump_sendfile) ERROR: ==== 2024-12-16T10:26:20 sendfile failed: Broken pipe
distccd[34098] (2024-12-16 10:26:20) (dcc_x_file) Warning: ==== 2024-12-16T10:26:20 Something went wrong with client recv file
distccd[34098] (2024-12-16 10:26:20) (dcc_x_token_int) send DOTO00000000
distccd[34098] (2024-12-16 10:26:20) (dcc_writex) ERROR: ==== 2024-12-16T10:26:20 failed to write: Broken pipe
distccd[34098] (2024-12-16 10:26:20) /home/relay/wuminghui03/workspace/all_svn/kmake/distcc-3.2.ks/clang++ .build/clang_opt/objs/teams/ad/ad_target/kconf_proto/utils/kconf/kconf_data.pb.o on localhost completed ok
distccd[34098] (2024-12-16 10:26:20) job complete
distccd[34098] (2024-12-16 10:26:20) (dcc_cleanup_tempfiles_inner) deleted 6 temporary files
distccd[34098] (2024-12-16 10:26:20) (dcc_job_summary) client: 10.106.192.98:1367 COMPILE_OK exit:0 sig:0 core:0 ret:0 time:305771ms /home/relay/wuminghui03/workspace/all_svn/kmake/distcc-3.2.ks/clang++ .build/clang_opt/objs/teams/ad/ad_target/kconf_proto/utils/kconf/kconf_data.pb.o
```
At this time, the client is still waiting for select, and it needs to wait for the timeout to end before local compilation.
I am currently using distcc v3.2, and I am not sure whether the latest version has solved it. If not, I can try to solve it myself.

**Comment by sourcefrog** [comment]
From memory there is no protocol-level heartbeat, just a TCP connection. The "broken pipe" here likely indicates that either the client process crashed, or there was some network connectivity error.

**Reply by zcfh** [reply, marked as answer]
After investigation, I found that it was related to the **network configuration of the container**.
Packet loss is more likely to occur when sending large files. After packet loss, the kernel will automatically modify a network configuration timeout. When the remote compilation time is long, the host will automatically disconnect TCP, but the TCP connection in the container is not aware of it.
I avoided this problem by turning on the TCP heartbeat mechanism. Perhaps this can be used as a switch to avoid similar problems. For example, the server sends a disconnection request for some reason, but the client does not perceive it and keeps waiting until the timeout.

**Reply by sourcefrog** [reply]
Maybe there's a sockopt we could use to turn on heartbeats
